Puzzle State Perf

tabascq opened the pull request:
Updating puzzle state on add/remove of puzzles/teams instead of
validating it all the time. I'm also validating on application start
for the time being, to help migrate existing events.

tabascq commented:
Eventually this PR may become self aware... :-)

morganbr commented on ServerCore/Pages/Teams/Play.cshtml:
```diff
@@ -0,0 +1,73 @@
+                @Html.DisplayFor(modelItem => item.Puzzle.PuzzleUrl)
+            </td>
+            <td>
+                @Html.DisplayFor(modelItem => item.Puzzle.AnswerUrl)
```
Is this an admin page or a player page? If it's a player page, maybe this should be conditional on the event's answers available time

tabascq replied:
It's a player page, hence the name "Play". I had a rationale at the time for skipping the conditional on answers, but I can't think of what that rationale could be at the moment. It was probably misguided at the time, will do.

morganbr commented on ServerCore/Pages/Teams/Play.cshtml.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace ServerCore.Pages.Teams
+{
+    public class PlayModel : EventSpecificPageModel
```
Could you please add an XML comment on what this view is?

morganbr commented on Data/DataModel/PuzzleStatePerTeam.cs:
```diff
@@ -68,5 +73,124 @@ public bool IsSolved
+
+                if (saveChanges)
+                {
+                    await context.SaveChangesAsync();
```
I think you might have handle it correctly in the code so far, but I'm a bit nervous about correct save handling. Is there a good way to encapsulate this to avoid errors?

tabascq replied:
Even a transaction won't work if one person adds a puzzle at (nearly) the same time that someone else adds a team. You've convinced me that I have no idea how to successfully do this, so I am removing all the perf-related parts of this change. The player page doesn't need to ensure that state exists, since it only needs to display puzzles that are unlocked and you can't unlock without having a record.

Any code that unlocks a puzzle via prerequisites will have to find its own way to ensure state. Eventually we should make a helper for this.

The status page will still need to enforce that state exists unless someone can show me how to create a join that includes missing items - I cannot figure out how to do this despite looking at several examples.

morganbr replied:
I'd suggest that we still have code that creates these when puzzles and teams are created. As long as it only calls ```SaveChangesAsync``` once, it will ensure there are no partially built states when there isn't a race condition.

tabascq replied:
That is exactly what the code you reviewed was doing. I am lost here.

tabascq replied:
Specifically, Create.cs passed saveChanges:false so that there would be a single SaveChangesAsync call. But you've convinced me that this isn't solid enough because a new team could have snuck in at the same time, so I don't know that it matters.

morganbr commented on Data/DataModel/PuzzleStatePerTeam.cs:
```diff
@@ -68,5 +73,124 @@ public bool IsSolved
         /// </summary>
         public string Notes { get; set; }
+
+        #region Consistency Helpers
```
This is all starting to be a lot of logic in the model. Should we consider somewhere else to put this kind of code?

tabascq replied:
Yep, will move to EventOperations as discussed before.

tabascq replied:
Nope, will remove it except for the subset that the Status page needs, which will move there.

tabascq commented on Data/DataModel/PuzzleStatePerTeam.cs:
```diff
@@ -68,5 +73,124 @@ public bool IsSolved
+        #region Consistency Helpers
+        public static async Task<bool> EnsureStateForPuzzleAsync(PuzzleServerContext context, int eventId, int puzzleId, bool saveChanges = true)
+        {
+            var teamsQ = context.Teams.Where(team => team.Event.ID == eventId).Select(team => team.ID);
```
I don't see how it will get left out - in your scenario there would be another Ensure method coming behind it for the new puzzle.

tabascq replied:
I'm pulling this part of the code. Now the status page will try to pull in state before it displays, and if it misses one in some unusual circumstance that's OK. It'll get it the next time.

tabascq commented on ServerCore/ModelBases/PuzzleStatePerTeamPageModel.cs:
```diff
@@ -0,0 +1,124 @@
+
+        public IList<PuzzleStatePerTeam> PuzzleStatePerTeam { get; set; }
+
+        protected abstract string DefaultSort { get; }
```
Ummm, maybe?

tabascq replied:
Answer: Yes and yes. Working on it, it's not too hard and I agree the type safety is good.

tabascq commented on Data/DataModel/PuzzleStatePerTeam.cs:
```diff
@@ -68,5 +73,124 @@ public bool IsSolved
+
+        public static async Task<bool> EnsureStateForEventAsync(PuzzleServerContext context, int eventId, bool saveChanges = true)
+        {
+            bool addedAnything = false;
```
I'm pulling this part of the code.

tabascq commented on ServerCore/Pages/Events/Delete.cshtml.cs:
```diff
@@ -39,6 +39,7 @@ public async Task<IActionResult> OnPostAsync(int id)
 
             if (Event != null)
             {
+                await PuzzleStatePerTeam.RemoveStateForEventAsync(_context, Event.ID, saveChanges: false);
```
I'm pulling this part of the code.

tabascq commented:
Hey @morganbr, I think I've addressed all your feedback. I don't think I've succeeded in the thing I set out to do, but I've added a lot of other things that are valuable and I'd like to get those in before taking another stab at state consistency, since I don't have a lot of confidence that I can address it quickly or properly and I don't want to hold the rest hostage. I'm done with the UX I wanted to light up this "sprint", so I will not let the next attempt undergo "feature creep" even if it takes a while to write and/or review.

Thanks for the feedback!

tabascq commented on ServerCore/ModelBases/PuzzleStatePerTeamPageModel.cs:
```diff
@@ -0,0 +1,180 @@
+            await Context.SaveChangesAsync();
+        }
+
+        public enum SortOrder
```
I would personally prefer it at the bottom, but some people get antsy about wanting private methods down there. If you prefer it at the bottom too, consider it done.
